logsmanagement/tail_plugin.c

```diff
@@ -0,0 +1,490 @@
+        m_assert(0, "Should never reach EOF");
+        goto free_access_lock;
+    }
+
+free_access_lock:
+    // debug(D_LOGS_MANAG, "Access_lock released for %s\n", p_file_info->file_basename);
+    (void)file_close(p_file_info);
+    (void)enable_file_changed_events(p_file_info, 1);
+    uv_fs_req_cleanup(req);
+}
+
+/**
+ * @brief Async callback function, called when a log file change event is fired.
+ * @param[in] handle Callback handle to pass data to the function.
+ * @param[in] file_basename Basename of the file that was changed.
+ * @param[in] events ORed mask of uv_fs_event elements, unused.
+ * @param[in] status Unused.
+ */
+static void file_changed_cb(uv_fs_event_t *handle, const char *file_basename, int events, int status) {
+    UNUSED(file_basename);
+    UNUSED(events);
+    UNUSED(status);
+
+    uv_fs_t stat_req;
+    uv_stat_t *statbuf;
+
+    int rc = 0;
+    struct File_info *p_file_info = handle->data;
+
+    // debug(D_LOGS_MANAG, "File changed! %s", file_basename ? file_basename : "");
+    // debug(D_LOGS_MANAG, "%s %s", file_basename, 
+    //         !p_file_info->force_file_changed_cb ? "forced file_changed_cb()" : "regular file_changed_cb()");
+
+    rc = uv_fs_event_stop(p_file_info->fs_event_req);
+    if (unlikely(rc)) {
+        error("uv_fs_event_stop() error for %s: %s", p_file_info->filename, uv_strerror(rc));
+        if (rc == UV_ENOENT) handle_UV_ENOENT_err(p_file_info);
+        else m_assert(!rc, "uv_fs_event_stop() failed");
+        goto return_error;
+    }
+
+    rc = uv_fs_stat(main_loop, &stat_req, p_file_info->filename, NULL);
+    if (unlikely(rc)) {
+        error("uv_fs_stat error: %s", uv_strerror(rc));
+        if (rc == UV_ENOENT) handle_UV_ENOENT_err(p_file_info);
+        else m_assert(!rc, "uv_fs_stat error");
+        goto return_error;
+    }
+    statbuf = uv_fs_get_statbuf(&stat_req);    
+    
+    if(unlikely(p_file_info->rotated)){
+        p_file_info->rotated = 0;
+        p_file_info->filename[strlen(p_file_info->filename) - 2] = '\0';
+        freez(p_file_info->file_basename);
+        p_file_info->file_basename = get_basename(p_file_info->filename);
+        // p_file_info->file_basename[strlen(p_file_info->file_basename - 2)] = '\0';
+        p_file_info->filesize = 0;
+
+        /* Need a new uv_fs_get_statbuf() call since filename changed */
+        uv_fs_req_cleanup(&stat_req);
+        rc = uv_fs_stat(main_loop, &stat_req, p_file_info->filename, NULL);
+        if (unlikely(rc)) {
+            error("uv_fs_stat error: %s\n", uv_strerror(rc));
+            m_assert(!rc, "uv_fs_stat error");
+            // TODO: Handle error in this case
+        }
+        statbuf = uv_fs_get_statbuf(&stat_req);
+
+    } else if(unlikely(p_file_info->inode != statbuf->st_ino)) {
+        char *new_filename = malloc(strlen(p_file_info->filename)+ 3);
+        new_filename[0] = '\0';
+        strcat(new_filename, p_file_info->filename);
```

Comment:
`strcat` is too expensive for the host, can you change for `strcpy`? Or to use either `string` or `buffer` here? :thinking:

Reply:
I don't think it matters too much in this particular case, as this code is run only when there is log rotation, which is very infrequent.